### PR TITLE
MULE-14963: Add triggers for Email connector

### PR DESCRIPTION
**Pull request by juandesi**

### CODE REVIEW

File: src/main/java/org/mule/extension/email/internal/EmailConnector.java [hunk 1]
```diff
@@ -27,8 +30,20 @@
...
+@Export(classes = {EmailException.class, com.sun.mail.handlers.multipart_mixed.class})
 public class EmailConnector {
 
   public static final String TLS_CONFIGURATION = "TLS Configuration";
+
+  @Parameter
+  @Optional
+  private multipart_mixed caca;
+
+  public multipart_mixed getCaca() {
```
Comment (diegostrubolini) [inline src/main/java/org/mule/extension/email/internal/EmailConnector.java]: 💩
this should be removed right?

File: src/test/munit/imap-trigger-test-case.xml [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+    <munit:after-suite name="shutDownServer-imap">
+        <java:invoke-static class="${imap.server}" method="stop()"/>
+    </munit:after-suite>
+
+    <!-- Configs -->
+
+    <email:imap-config name="imap-config">
+        <email:imap-connection host="127.0.0.1" port="${imap.port}" user="juan.desimoni@mulesoft.com" password="password">
+            <reconnection failsDeployment="false">
```
Comment (diegostrubolini) [inline src/test/munit/imap-trigger-test-case.xml]: is this still necessary?

File: src/test/munit/imap-trigger-test-case.xml [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+                <munit-tools:assert-that expression="#[attributes.subject]" is="#[MunitTools::equalTo('Find me matcher')]"/>
+            </foreach>
+        </munit:validation>
+
+    </munit:test>
+
+    <flow name="trigger-with-matcher-imap" maxConcurrency="1">
+
+        <email:on-email-imap config-ref="imap-config" >
+            <reconnect-forever blocking="false"/>
```
Comment (diegostrubolini) [inline src/test/munit/imap-trigger-test-case.xml]: same here

File: src/test/java/org/mule/extension/email/EmailConnectorTestCase.java [hunk 4]
```diff
@@ -37,6 +37,7 @@
 import org.junit.Rule;
 
 @ArtifactClassLoaderRunnerConfig(applicationSharedRuntimeLibs = {"com.sun.mail:javax.mail"},
+    testExclusions = {"org.mule.module:mule-java-module"},
```
Comment (diegostrubolini) [inline src/test/java/org/mule/extension/email/EmailConnectorTestCase.java]: I think they fixed this

Comment (juandesi) [reply src/test/java/org/mule/extension/email/EmailConnectorTestCase.java]: they fixed for another version

File: src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import javax.mail.UIDFolder;
+
+import static java.util.Optional.ofNullable;
+
+/**
+ * Retrieves all the emails from an IMAP mailbox folder, watermark can be enabled for polled items.
+ *
+ * @since 1.1
+ */
+@DisplayName("On Email - IMAP")
```
Comment (estebanwasinger) [inline src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java]: On Email -> On New Email?

File: src/main/java/org/mule/extension/email/internal/mailbox/pop3/POP3PollingSource.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import javax.mail.Message;
+
+import static java.util.Optional.ofNullable;
+
+/**
+ * Retrieves all the emails from an POP3 mailbox folder.
+ *
+ * @since 1.1
+ */
+@DisplayName("On Email - POP3")
```
Comment (estebanwasinger) [inline src/main/java/org/mule/extension/email/internal/mailbox/pop3/POP3PollingSource.java]: same here

Comment (juandesi) [reply src/main/java/org/mule/extension/email/internal/mailbox/pop3/POP3PollingSource.java]: but not always are new, this case probably most of the emails are not new

File: src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java (hunk 5 above)
Comment (juandesi) [reply src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java]: but not always are new

Comment (estebanwasinger) [reply src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java]: yes, but is like DB, the first iteration won't be new emails, but if you keep the source alive, will be always new emails, right?

File: src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    }
+  }
+
+  /**
+   * Marks an email as deleted looking it by its UID.
+   */
+  private void markAsDeleted(String id) {
+    try {
+      if (openFolder instanceof POP3Folder) {
+        for (Message message : openFolder.getMessages()) {
```
Comment (alegmarra) [inline src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java]: you already had all the messages from the poll, isn't this costly to obtain twice?

Comment (alegmarra) [reply src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java]: in any case, use the `getMessages(openFolder)` like before

File: src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+          if (((POP3Folder) openFolder).getUID(message).equals(id)) {
+            message.setFlag(DELETED, true);
```
Comment (alegmarra) [inline src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java]: break?

File: src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+        message.setFlag(DELETED, true);
+      }
+    } catch (MessagingException e) {
+      throw new ExpungeFolderException("Error while setting delete flag on email uid [" + id + "]", e);
+    }
+  }
+
+  private DefaultStoredEmailContent getEmailContent(Message message, String id) {
+    try {
+      return (DefaultStoredEmailContent) storedEmailContentFactory.fromMessage(message);
```
Comment (alegmarra) [inline src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java]: why the downcast?

File: src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+@Alias("on-email-imap")
```
Comment (alegmarra) [inline src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java]: imap-listener

File: src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  @DisplayName("Enable Watermark")
+  @Optional(defaultValue = "false")
+  private boolean watermarkEnabled;
+
+  /**
+   * A matcher to filter emails retrieved by this polling source.
+   */
+  @Parameter
+  @Optional
+  private IMAPEmailPredicateBuilder imapMatcher;
```
Comment (alegmarra) [inline src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java]: `@NullSafe(defaultImplementingType = UnreadEmailMatcher.class)`

File: src/main/java/org/mule/extension/email/internal/mailbox/pop3/POP3PollingSource.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+@Alias("on-email-pop3")
```
Comment (alegmarra) [inline src/main/java/org/mule/extension/email/internal/mailbox/pop3/POP3PollingSource.java]: pop3-listener

File: src/test/munit/imap-secure-trigger-test-case.xml [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+    <sub-flow name="storeMessages-imaps">
+        <set-variable variableName="message" value="#[message]"/>
+        <munit-tools:retrieve key="messages"/>
+        <munit-tools:store key="messages">
+            <munit-tools:value>
+                #[
+                %dw 2.0
+                output application/java
+                ---
+                payload ++ [vars.message]
```
Comment (alegmarra) [inline src/test/munit/imap-secure-trigger-test-case.xml]: why, oh why, not use the mtf:queue?

File: src/test/munit/imap-secure-trigger-test-case.xml [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+            </java:args>
+        </java:invoke-static>
+    </munit:before-suite>
+
+    <munit:before-test name="populate-mailbox-imaps">
+        <foreach collection="#[1 to 10]">
+            <java:invoke-static class="${imap.server}" method="sendEmail()"/>
+        </foreach>
+        <munit-tools:store key="messages">
+            <munit-tools:value>#[[]]</munit-tools:value>
```
Comment (alegmarra) [inline src/test/munit/imap-secure-trigger-test-case.xml]: sounds like mtf:queue

File: src/test/munit/imap-secure-trigger-test-case.xml [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+    <munit:test name="check-messages-content-imaps">
+
+        <munit:enable-flow-sources>
+            <munit:enable-flow-source value="trigger-simple-imaps"/>
+        </munit:enable-flow-sources>
+
+        <munit:execution>
+            <until-successful maxRetries="10" millisBetweenRetries="500">
+                <munit-tools:retrieve key="messages"/>
+                <validation:validate-size value="#[payload]" min="10" max="10"/>
```
Comment (alegmarra) [inline src/test/munit/imap-secure-trigger-test-case.xml]: `<munit-tools:assert-that expression="#[Mtf::queueSize()]" is="#[MunitTools::equalTo(10)]"/>`?

File: src/test/munit/imap-secure-trigger-test-case.xml [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+    <!-- Test polled emails are being filtered with a Matcher -->
+    <munit:test name="with-matcher-imaps">
+
+        <munit:enable-flow-sources>
+            <munit:enable-flow-source value="trigger-with-matcher-imaps"/>
+        </munit:enable-flow-sources>
+
+        <munit:behavior>
+            <foreach collection="#[1 to 2]">
+                <java:invoke-static class="${imap.server}" method="sendEmailWithSubject(String, String)">
```
Comment (alegmarra) [inline src/test/munit/imap-secure-trigger-test-case.xml]: why not using the `email:send` operation?

File: src/test/munit/imap-secure-trigger-test-case.xml [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+        </email:on-email-imap>
+
+        <flow-ref name="storeMessages-imaps"/>
+
+    </flow>
+
+    <!-- Tests polled emails are watermarked and new emails are delivered -->
+    <!-- TODO: IGNORED until this can be run in a specific mule version -->
+    <!-- TODO: IGNORED until this can be run in a specific mule version -->
+    <!-- TODO: IGNORED until this can be run in a specific mule version -->
```
Comment (alegmarra) [inline src/test/munit/imap-secure-trigger-test-case.xml]: ticket please

File: pom.xml [hunk 18]
```diff
@@ -27,13 +27,96 @@
...
+                    <execution>
+                        <goals>
+                            <goal>test</goal>
+                        </goals>
+                        <phase>integration-test</phase>
+                    </execution>
+                </executions>
+                <configuration>
+                    <targetRuntimes>
+                        <targetRuntime>MULE:4.1.1</targetRuntime>
```
[the remark this answers is not in the document]
Comment (juandesi) [reply pom.xml]: It wont work, autodiscovery only works when you don't specify rts

File: src/main/java/org/mule/extension/email/internal/EmailConnector.java [hunk 19]
```diff
@@ -27,8 +30,20 @@
 @Configurations({SMTPConfiguration.class, POP3Configuration.class, IMAPConfiguration.class})
 @Extension(name = "Email")
 @ErrorTypes(EmailError.class)
-@Export(classes = EmailException.class)
+@Export(classes = {EmailException.class, com.sun.mail.handlers.multipart_mixed.class})
```
[the remark this answers is not in the document]
Comment (juandesi) [reply src/main/java/org/mule/extension/email/internal/EmailConnector.java]: Workaround that I forgot to delete, LOL

File: pom.xml (hunk 18 above)
[the remark this answers is not in the document]
Comment (diegostrubolini) [reply pom.xml]: they should be complimentary, not mutually exclusive 😀

File: src/main/java/org/mule/extension/email/internal/StoredEmailContentFactory.java [hunk 20]
```diff
@@ -143,7 +145,8 @@ private void processBodyPart(Part part, StringJoiner bodyCollector, Map<String,
    */
   private void processAttachment(Part part, Map<String, TypedValue<InputStream>> attachments, StreamingHelper streamingHelper)
       throws MessagingException, IOException {
-    Object content = streamingHelper.resolveCursorProvider(part.getInputStream());
+    Object content =
+        streamingHelper != null ? streamingHelper.resolveCursorProvider(part.getInputStream()) : part.getInputStream();
```
[the remark this answers is not in the document]
Comment (juandesi) [reply src/main/java/org/mule/extension/email/internal/StoredEmailContentFactory.java]: sources, the stream is wrapped once pushed to the flow tho

File: src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+  private MailboxConnection connection;
+
+  private Predicate<BaseEmailAttributes> predicate;
+
+  /**
+   * The name of the folder to poll emails from. Defaults to "INBOX".
+   */
+  @Parameter
+  @Optional(defaultValue = "INBOX")
+  private String folder;
```
[the remark this answers is not in the document]
Comment (juandesi) [reply src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java]: not for now, could be a good improvement tho.

[the remark this answers is not in the document]
Comment (juandesi) [reply src/main/java/org/mule/extension/email/internal/mailbox/BaseMailboxPollingSource.java]: https://www.mulesoft.org/jira/browse/MULE-15430

File: src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java (hunk 11 above)
Comment (juandesi) [reply src/main/java/org/mule/extension/email/internal/mailbox/imap/IMAPPollingSource.java]: Can't use that on a concrete type

File: pom.xml (hunk 18 above)
[the remark this answers is not in the document]
Comment (juandesi) [reply pom.xml]: oks, I still think it's better to explicitly put all the RTS tho.

**Pull request merged**